src/Update/Nix/FetchGit.hs

```diff
@@ -128,48 +72,44 @@ extractFetchGitArgs = \case
-fetchGitArgsToUpdate t as = FetchGitUpdateInfo (extractUrlString (repoLocation as))
-                                           <$> exprSpan t (revExpr as)
-                                           <*> exprSpan t (sha256Expr as)
+getFetchGitLatestInfo :: FetchGitArgs -> IO (Either Warning FetchGitLatestInfo)
+getFetchGitLatestInfo args = do
+  result <- nixPrefetchGit (extractUrlString $ repoLocation args)
+  case result of
+    Left e -> pure $ Left e
+    Right o -> return $ FetchGitLatestInfo args (rev o) (sha256 o)
+                                           <$> (parseISO8601DateToDay (date o))
 
 --------------------------------------------------------------------------------
--- The type we have after calling nix-prefetch-git, A pair of SpanUpdates for
--- each fetchgit call.
+-- Deciding which parts of the Nix file should be updated and how.
 --------------------------------------------------------------------------------
 
--- | A pair of 'SpanUpdate's for updating a single fetchgit value.
-data FetchGitSpanUpdates = FetchGitSpanUpdates{ revUpdate    :: SpanUpdate
-                                              , sha256Update :: SpanUpdate
-                                              }
-  deriving (Show, Data)
-
--- | Create a pair of 'SpanUpdate's given a 'FetchGitUpdateInfo'
-updateInfoToUpdate :: FetchGitUpdateInfo -> IO (Either Warning FetchGitSpanUpdates)
-updateInfoToUpdate ui = do
-  o <- nixPrefetchGit (urlString ui)
-  pure $ prefetchGitOutputToSpanUpdate ui <$> o
-
--- | Use the output of nix-prefetch-git to create a pair of updates
-prefetchGitOutputToSpanUpdate :: FetchGitUpdateInfo
-                              -> NixPrefetchGitOutput
-                              -> FetchGitSpanUpdates
-prefetchGitOutputToSpanUpdate u o =
-  FetchGitSpanUpdates (SpanUpdate (revPos u) (quote $ rev o))
-                      (SpanUpdate (sha256Pos u) (quote $ sha256 o))
-  where quote t = "\"" <> t <> "\""
-
-
+fetchTreeToSpanUpdates :: FetchTree FetchGitLatestInfo -> [SpanUpdate]
+fetchTreeToSpanUpdates (Node maybeVersionExpr cs) =
+  (concatMap fetchTreeToSpanUpdates cs)
+  ++ (toList $ maybeVersionExpr >>= maybeUpdateVersion cs)
+fetchTreeToSpanUpdates (FetchNode f) = [revUpdate, sha256Update]
+  where revUpdate = SpanUpdate (exprSpan (revExpr args))
+                               (quoteString (latestRev f))
+        sha256Update = SpanUpdate (exprSpan (sha256Expr args))
+                                  (quoteString (latestSha256 f))
+        args = (originalInfo f)
+
+-- Given a Nix expression representing a version value, and the
+-- children of the node that contains it, decides whether and how it
+-- should that version string should be updated.  We basically just
+-- take the latest date of all the fetches in the children.
+maybeUpdateVersion :: [FetchTree FetchGitLatestInfo] -> NExprLoc
+                   -> Maybe SpanUpdate
+maybeUpdateVersion cs versionExpr =
+  case versionDays (Node Nothing cs) of
```

Comment:
It seems a little odd to construct a `Node` here just to pass to `versionDays`. An alternative might be to use `universeBi` to get all the `FetchNode` values and take the maximum day from those.